**复习：**在前面我们已经学习了Pandas基础，第二章我们开始进入数据分析的业务部分，在第二章第一节的内容中，我们学习了**数据的清洗**，这一部分十分重要，只有数据变得相对干净，我们之后对数据的分析才可以更有力。而这一节，我们要做的是数据重构，数据重构依旧属于数据理解（准备）的范围。

#### 开始之前，导入numpy、pandas包和数据

In [1]:
import numpy as np
import pandas as pd

In [2]:
# 载入data文件中的:train-left-up.csv
text = pd.read_csv('data/train-left-up.csv')
text.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  
0                            Braund, Mr. Owen Harris  
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  
2                             Heikkinen, Miss. Laina  
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  
4                           Allen, Mr. William Henry

# 2 第二章：数据重构


### 2.4 数据的合并

#### 2.4.1 任务一：将data文件夹里面的所有数据都载入，观察数据的之间的关系

In [5]:
text_left_up = pd.read_csv("data/train-left-up.csv")
text_left_down = pd.read_csv("data/train-left-down.csv")
text_right_up = pd.read_csv("data/train-right-up.csv")
text_right_down = pd.read_csv("data/train-right-down.csv")
text_left_up.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  
0                            Braund, Mr. Owen Harris  
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  
2                             Heikkinen, Miss. Laina  
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  
4                           Allen, Mr. William Henry

In [13]:
text_left_down.head()

PassengerId  Survived  Pclass                                         Name
0          440         0       2       Kvillner, Mr. Johan Henrik Johannesson
1          441         1       2  Hart, Mrs. Benjamin (Esther Ada Bloomfield)
2          442         0       3                              Hampe, Mr. Leon
3          443         0       3                    Petterson, Mr. Johan Emil
4          444         1       2                    Reynaldo, Ms. Encarnacion

In [11]:
text_right_up

Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked
0      male  22.0    1.0    0.0         A/5 21171   7.2500   NaN        S
1    female  38.0    1.0    0.0          PC 17599  71.2833   C85        C
2    female  26.0    0.0    0.0  STON/O2. 3101282   7.9250   NaN        S
3    female  35.0    1.0    0.0            113803  53.1000  C123        S
4      male  35.0    0.0    0.0            373450   8.0500   NaN        S
..      ...   ...    ...    ...               ...      ...   ...      ...
886    male  27.0    0.0    0.0            211536  13.0000   NaN        S
887  female  19.0    0.0    0.0            112053  30.0000   B42        S
888  female   NaN    1.0    2.0        W./C. 6607  23.4500   NaN        S
889    male  26.0    0.0    0.0            111369  30.0000  C148        C
890    male  32.0    0.0    0.0            370376   7.7500   NaN        Q

[891 rows x 8 columns]

In [14]:
text_right_down.head()

Sex   Age  SibSp  Parch        Ticket    Fare Cabin Embarked
0    male  31.0      0      0    C.A. 18723  10.500   NaN        S
1  female  45.0      1      1  F.C.C. 13529  26.250   NaN        S
2    male  20.0      0      0        345769   9.500   NaN        S
3    male  25.0      1      0        347076   7.775   NaN        S
4  female  28.0      0      0        230434  13.000   NaN        S

【提示】结合之前我们加载的train.csv数据，大致预测一下上面的数据是什么

<div class="alert alert-block alert-success">
【回答】以上四个文件都是从train.csv中截取的部分。其中left是指数据框的左边部分，包括PassengerId、Survived、Pclass、Name四列；right则包括Sex、Age、SibSp、Parch、Ticket、Fare、Cabin、Embarked八列信息。up指表格从头到尾，而down是从440行开始截取表格。暂时不知道数据的用途</div>

#### 2.4.2：任务二：使用concat方法：将数据train-left-up.csv和train-right-up.csv横向合并为一张表，并保存这张表为result_up

In [43]:
list_up = [text_left_up,text_right_up]
result_up = pd.concat(list_up,axis=1)
result_up

PassengerId  Survived  Pclass  \
0            1.0       0.0     3.0   
1            2.0       1.0     1.0   
2            3.0       1.0     3.0   
3            4.0       1.0     1.0   
4            5.0       0.0     3.0   
..           ...       ...     ...   
886          NaN       NaN     NaN   
887          NaN       NaN     NaN   
888          NaN       NaN     NaN   
889          NaN       NaN     NaN   
890          NaN       NaN     NaN   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0    1.0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                               Heikkinen, Miss. Laina  female  26.0    0.0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                             Allen, Mr. William Henry    male  35.0    0.0   
..                                                 ...     ...   ...    ...   
886                                                NaN    male  27.0    0.0   
887                                                NaN  female  19.0    0.0   
888                                                NaN  female   NaN    1.0   
889                                                NaN    male  26.0    0.0   
890                                                NaN    male  32.0    0.0   

     Parch            Ticket     Fare Cabin Embarked  
0      0.0         A/5 21171   7.2500   NaN        S  
1      0.0          PC 17599  71.2833   C85        C  
2      0.0  STON/O2. 3101282   7.9250   NaN        S  
3      0.0            113803  53.1000  C123        S  
4      0.0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886    0.0            211536  13.0000   NaN        S  
887    0.0            112053  30.0000   B42        S  
888    2.0        W./C. 6607  23.4500   NaN        S  
889    0.0            111369  30.0000  C148        C  
890    0.0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

#### 2.4.3 任务三：使用concat方法：将train-left-down和train-right-down横向合并为一张表，并保存这张表为result_down。然后将上边的result_up和result_down纵向合并为result。

In [44]:
list_down=[text_left_down,text_right_down]
result_down = pd.concat(list_down,axis=1)
result = pd.concat([result_up,result_down])
result

PassengerId  Survived  Pclass  \
0            1.0       0.0     3.0   
1            2.0       1.0     1.0   
2            3.0       1.0     3.0   
3            4.0       1.0     1.0   
4            5.0       0.0     3.0   
..           ...       ...     ...   
447        887.0       0.0     2.0   
448        888.0       1.0     1.0   
449        889.0       0.0     3.0   
450        890.0       1.0     1.0   
451        891.0       0.0     3.0   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0    1.0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                               Heikkinen, Miss. Laina  female  26.0    0.0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                             Allen, Mr. William Henry    male  35.0    0.0   
..                                                 ...     ...   ...    ...   
447                              Montvila, Rev. Juozas    male  27.0    0.0   
448                       Graham, Miss. Margaret Edith  female  19.0    0.0   
449           Johnston, Miss. Catherine Helen "Carrie"  female   NaN    1.0   
450                              Behr, Mr. Karl Howell    male  26.0    0.0   
451                                Dooley, Mr. Patrick    male  32.0    0.0   

     Parch            Ticket     Fare Cabin Embarked  
0      0.0         A/5 21171   7.2500   NaN        S  
1      0.0          PC 17599  71.2833   C85        C  
2      0.0  STON/O2. 3101282   7.9250   NaN        S  
3      0.0            113803  53.1000  C123        S  
4      0.0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
447    0.0            211536  13.0000   NaN        S  
448    0.0            112053  30.0000   B42        S  
449    2.0        W./C. 6607  23.4500   NaN        S  
450    0.0            111369  30.0000  C148        C  
451    0.0            370376   7.7500   NaN        Q  

[1343 rows x 12 columns]

#### 2.4.4 任务四：使用DataFrame自带的方法join方法和append：完成任务二和任务三的任务

In [49]:
result_up = text_left_up.join(text_right_up)
result_down = text_left_down.join(text_right_down)
result = result_up.append(result_down)
result

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
434          435         0       1   
435          436         1       1   
436          437         0       3   
437          438         1       2   
438          439         0       1   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0    1.0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                               Heikkinen, Miss. Laina  female  26.0    0.0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                             Allen, Mr. William Henry    male  35.0    0.0   
..                                                 ...     ...   ...    ...   
434                          Silvey, Mr. William Baird    male  50.0    1.0   
435                          Carter, Miss. Lucile Polk  female  14.0    1.0   
436               Ford, Miss. Doolina Margaret "Daisy"  female  21.0    2.0   
437              Richards, Mrs. Sidney (Emily Hocking)  female  24.0    2.0   
438                                  Fortune, Mr. Mark    male  64.0    1.0   

     Parch            Ticket      Fare        Cabin Embarked  
0      0.0         A/5 21171    7.2500          NaN        S  
1      0.0          PC 17599   71.2833          C85        C  
2      0.0  STON/O2. 3101282    7.9250          NaN        S  
3      0.0            113803   53.1000         C123        S  
4      0.0            373450    8.0500          NaN        S  
..     ...               ...       ...          ...      ...  
434    0.0             13507   55.9000          E44        S  
435    2.0            113760  120.0000      B96 B98        S  
436    2.0        W./C. 6608   34.3750          NaN        S  
437    3.0             29106   18.7500          NaN        S  
438    4.0             19950  263.0000  C23 C25 C27        S  

[439 rows x 12 columns]

#### 2.4.5 任务五：使用Panads的merge方法和DataFrame的append方法：完成任务二和任务三的任务

In [7]:
result_up = pd.merge(text_left_up,text_right_up,left_index=True,right_index=True)
result_down = pd.merge(text_left_down,text_right_down,left_index=True,right_index=True)
result = result_up.append(result_down)
result

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
447          887         0       2   
448          888         1       1   
449          889         0       3   
450          890         1       1   
451          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0    1.0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                               Heikkinen, Miss. Laina  female  26.0    0.0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                             Allen, Mr. William Henry    male  35.0    0.0   
..                                                 ...     ...   ...    ...   
447                              Montvila, Rev. Juozas    male  27.0    0.0   
448                       Graham, Miss. Margaret Edith  female  19.0    0.0   
449           Johnston, Miss. Catherine Helen "Carrie"  female   NaN    1.0   
450                              Behr, Mr. Karl Howell    male  26.0    0.0   
451                                Dooley, Mr. Patrick    male  32.0    0.0   

     Parch            Ticket     Fare Cabin Embarked  
0      0.0         A/5 21171   7.2500   NaN        S  
1      0.0          PC 17599  71.2833   C85        C  
2      0.0  STON/O2. 3101282   7.9250   NaN        S  
3      0.0            113803  53.1000  C123        S  
4      0.0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
447    0.0            211536  13.0000   NaN        S  
448    0.0            112053  30.0000   B42        S  
449    2.0        W./C. 6607  23.4500   NaN        S  
450    0.0            111369  30.0000  C148        C  
451    0.0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

【思考】对比merge、join以及concat的方法的不同以及相同。思考一下在任务四和任务五的情况下，为什么都要求使用DataFrame的append方法，如何只要求使用merge或者join可不可以完成任务四和任务五呢？

<div class="alert alert-block alert-success">
【回答】concat不管列名，axis = 0是增加行数，axis = 1时增加列数。merge和join都是用来连接表格，横向增加需要借助append方法。 </div>

#### 2.4.6 任务六：完成的数据保存为result.csv

In [8]:
result.to_csv('result.csv')

### 2.5 换一种角度看数据

#### 2.5.1 任务一：将我们的数据变为Series类型的数据

In [11]:
# 重新加载完整数据
text = pd.read_csv('result.csv')
unit_result=text.stack().head(20)
#type(unit_result)
unit_result.head

<bound method NDFrame.head of 0  Unnamed: 0                                                     0
   PassengerId                                                    1
   Survived                                                       0
   Pclass                                                         3
   Name                                     Braund, Mr. Owen Harris
   Sex                                                         male
   Age                                                           22
   SibSp                                                          1
   Parch                                                          0
   Ticket                                                 A/5 21171
   Fare                                                        7.25
   Embarked                                                       S
1  Unnamed: 0                                                     1
   PassengerId                                                    2
   Survived       

<div class="alert alert-block alert-success">
stack的意思是堆叠，堆积，即树状，与之对应的时unstack，数据则会以表格形式展示。可参见博客：https://www.cnblogs.com/bambipai/p/7658311.html</div>

### 第二部分：数据聚合与运算

### 2.6 数据运用

#### 2.6.1 任务一：通过教材《Python for Data Analysis》P303、Google or anything来学习了解GroupBy机制

<div class="alert alert-block alert-success">
顾名思义，groupby就是更具一定条件对数据集进行分组，以进行后续的统计。 </div>


#### 2.6.2：任务二：计算泰坦尼克号男性与女性的平均票价

In [19]:
text['Fare'].groupby(text['Sex']).mean()

Sex
female    44.479818
male      25.523893
Name: Fare, dtype: float64

#### 2.6.3：任务三：统计泰坦尼克号中男女的存活人数

In [14]:
text['Survived'].groupby(text['Sex']).sum()

Sex
female    233
male      109
Name: Survived, dtype: int64

#### 2.6.4：任务四：计算客舱不同等级的存活人数

In [23]:
text['Survived'].groupby(text['Pclass']).sum()
# text['Survived'].groupby(text['Pclass']).sum()/text['Unnamed: 0'].groupby(text['Pclass']).count()

Pclass
1    0.629630
2    0.472826
3    0.242363
dtype: float64

【**思考**】从数据分析的角度，上面的统计结果可以得出那些结论

<div class="alert alert-block alert-success">
初步可以看出不同性别群体在票价有所差别，或许是由于男性整体上要低于女性，具体原因有待进一步分析。<br>
性别和客舱等级对存活率也有一定影响，一定程度上反映了对女性的关注，而不同客舱的存活人数有区别，但三种等级的存活率分别为63.0%、47.3%和24.2%，但具体影响是否显著需要进一步建模分析。</div>


【思考】从任务二到任务四中，这些运算可以通过agg()函数来同时计算。并且可以使用rename函数修改列名。你可以按照提示写出这个过程吗？

In [12]:
#text.groupby('Survived').agg({'Sex': 'mean', 'Pclass': 'count'}).rename(columns= {'Sex': 'mean_sex', 'Pclass': 'count_pclass'})
#df.groupby('Survived').agg({'Sex': 'mean', 'Pclass': 'count'})

text.groupby('Pclass').agg({'Age': 'mean', 'Survived': 'sum'}).rename(columns= {'Age': 'mean_age', 'Survived': 'sum_Survived'})

mean_age  sum_Survived
Pclass                         
1       38.233441           136
2       29.877630            87
3       25.140620           119

#### 2.6.5：任务五：统计在不同等级的票中的不同年龄的船票花费的平均值

In [32]:
text.groupby(['Pclass','Age'])['Fare'].mean().head()

Pclass  Age  
1       0.92     151.5500
        2.00     151.5500
        4.00      81.8583
        11.00    120.0000
        14.00    120.0000
Name: Fare, dtype: float64

#### 2.6.6：任务六：将任务二和任务三的数据合并，并保存到sex_fare_survived.csv

In [36]:
means  = text['Fare'].groupby(text['Sex']).mean()
survived_pclass = text['Survived'].groupby(text['Pclass']).sum()
result = pd.merge(means,survived_sex,on='Sex')
result

Fare  Survived
Sex                        
female  44.479818       233
male    25.523893       109

In [ ]:
result.to_csv('sex_fare_survived.csv')

#### 2.6.7：任务七：得出不同年龄的总的存活人数，然后找出存活人数的最高的年龄，最后计算存活人数最高的存活率（存活人数/总人数）


In [37]:
#不同年龄的存活人数
survived_age = text['Survived'].groupby(text['Age']).sum()
survived_age.head()

Age
0.42    1
0.67    1
0.75    2
0.83    2
0.92    1
Name: Survived, dtype: int64

In [38]:
#找出最大值的年龄段
survived_age[survived_age.values==survived_age.max()]

Age
24.0    15
Name: Survived, dtype: int64

In [39]:
_sum = text['Survived'].sum()
print(_sum)

342


In [40]:
#首先计算总人数
_sum = text['Survived'].sum()

print("sum of person:"+str(_sum))

precetn =survived_age.max()/_sum

print("最大存活率："+str(precetn))

sum of person:342
最大存活率：0.043859649122807015
